### PR TITLE
Haproxy to microservice

### PR DESCRIPTION
**Pull request by alena1108**
https://github.com/rancher/rancher/issues/2179

### CODE REVIEW

**Comment by blaggacao** [comment]
Would you mind giving encuraging UI options for https://github.com/rancher/rancher/issues/2179#issuecomment-203178448 ?

**Comment by vincent99** [comment]
-1 from me.. The point of this is to make it possible to put in whatever config you want, while preserving 1st-class options for the common use cases.  None of those are really commonly used (attaining arbitrary scores on SSL test notwithstanding), and most of these could/should be set by the application.  HSTS and HPKP would be the most relevant and I could see an argument for those since one of the 1st class features is SSL termination.. but they are also particularly dangerous and can lock users out en-masse for `max-age` if misconfigured.

**Comment by blaggacao** [comment]
That's indeed a pack of diligent arguments.

**Comment by blaggacao** [comment]
I wanted to draw another use case, which is more closely related to the rancher stack (or might be in future): A lot of Catalogue stacks out there have poorly secured api endpoints or web client interfaces. Example: Graylog input api endpoint (isn't yet a catalogue item, I'm on it) needs to be exposed publicly for docker daemons being able to talk to them and there is not one grain of protection on it. Vendor suggests reverse proxy. My imagination is just about to go wild what would be possible with rancher's user management exposed to the load balancers or any kind of client side certificate integrations... :smile:

**Comment by vincent99** [comment]
I'm not familiar with graylog in detail but I don't see why you'd have to publish it to a host port.. Any port the container is listening on is available on the 10.42.x.y address/DNS name of the container from inside the managed network, they don't need to be published to a host port to be used.

Using rancher's authentication to authorize a HTTP request would probably better suited to a nginx-based balancer where you can do lua scripting to talk to the Rancher API directly.  I don't think you can do much useful with what haproxy has built-in (an essentially hardcoded `userlist` with plaintext passwords, or hashed with algorithms we don't use for local auth.  In either case have no way to create that for any of the other external forms of auth).

**Comment by blaggacao** [comment]
I think the point here is that it is the docker daemon's logging driver which talks, and it just failed on that locally, so I assume there is no way to make the docker daemon itself (!) join the managed network.

I'm eagerly looking forward to witness advances in this field as they unfold. This PR is very exciting! As I understood it from [here](https://github.com/rancher/rancher/issues/4443#issuecomment-236801789) putting nginx in place was an afterthaught. :+1:

**Comment by blaggacao** [comment]
Have you any stance on GORB so far?
As I figured out, IoT applications are stuck on Rancher due to the lack of UDP support as for now.
https://github.com/kobolog/gorb
https://blog.codeship.com/kernel-load-balancing-for-docker-containers-using-ipvs/

File: provider/haproxy/haproxy_config_util.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+	global["ssl-default-server-ciphers "] = "ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:kEDH+AESGCM:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:ECDHE-ECDSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES128-SHA:DHE-DSS-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-DSS-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-RSA-DES-CBC3-SHA:ECDHE-ECDSA-DES-CBC3-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:AES:CAMELLIA:DES-CBC3-SHA:!aNULL:!eNULL:!EXPORT:!DES:!RC4:!MD5:!PSK:!aECDH:!EDH-DSS-DES-CBC3-SHA:!EDH-RSA-DES-CBC3-SHA:!KRB5-DES-CBC3-SHA"
+	global["ssl-default-bind-options "] = "no-sslv3 no-tlsv10"
+	global["chroot "] = "/var/lib/haproxy"
+	global["user  "] = "haproxy"
+	global["group "] = "haproxy"
+	global["daemon "] = ""
+	global["log 127.0.0.1:8514 local0 "] = ""
+	global["log 127.0.0.1:8514 local1 "] = "notice"
+
+	defaults["mode "] = "tcp"
```
Comment (ibuildthecloud) [inline provider/haproxy/haproxy_config_util.go]: Why spaces in all these keys?

**Pull request merged**